freedomofpress/securedrop
# Rework Ansible restore role to detect Onion service version mismatch

## Status

Ready for review

## Description of Changes

Make the restore playbook extract the Tor configuration from the backup, retrieve the app server's Tor configuration, and compare them. If the service versions differ, abort the restore with an explanation.

Fixes #4786.

## Testing

- Install the latest 1.0.0 release candidate on hardware or VMs, specifying only `v2_onion_services` with `sdconfig`/`site-specific`.
- Check out this branch on your admin workstation.
- Create a marker file under `/var/lib/securedrop/store` to test the restore. Send a submission through the source interface, or simply create a file manually. If you've got the time to try a largish file, great.
- Run `securedrop-admin backup` to backup the v2-only system.
- Delete the marker file.
- Run `securedrop-admin restore` to restore that backup. Confirm that the restore worked by verifying the presence of the marker file.
- Add v3 with `sdconfig`, then `securedrop-admin install`, then `securedrop-admin tailsconfig`.
- Run `securedrop-admin restore` with the backup file created earlier. The restore should abort with this message:
```
The Tor configuration on the app server offers version 2 and 3 services.
The Tor configuration in this backup offers version 2 services.

Restoring a backup with a different Tor configuration than the server 
is currently unsupported. If you require technical assistance, please 
contact the SecureDrop team via the support portal or at 
securedrop@freedom.press.
```

## Deployment

This reworks the restore playbook, replacing the `restore.py` script previously copied to the server with equivalent Ansible tasks, preventing hangs when Tor is restarted on the server. Those changes are straightforward but deserve scrutiny.

## Checklist

### If you made changes to `securedrop-admin`:


### If you made non-trivial code changes:
